File: PROB1.java
//TC:O(logn)
//SC:O(1)

class Solution {
    public double myPow(double x, int n) 
    {
        
       if(n==0)
           return 1;
       else
       {
         double val = myPow(x,n/2);  
         if(n%2==0)  
           return val*val;
         else
         {
             if(n<0)
               return val*val*1/x;
             else
               return val*val*x;
             
             
         }
           
           
           
           
           
       }
        
        
        
        
        
    }
}

File: prob1.py
# Time Complexity :O(logn)
# Space Complexity :O(n)
# Leet Code: Yes

class Solution:
    def myPow(self, x: float, n: int) -> float:
        def helper(x, n):
            # base case
            if n == 0:
                return 1

            # logic
            if n < 0:
                return helper(1 / x, -n)
            if n % 2 == 0:
                return helper(x * x, n // 2)
            else:
                return x * helper(x * x, n // 2)

        return helper(x, n)